File: fingerprints.py
import sys
import os
import numpy as np
from rdkit import Chem
from rdkit.Chem import rdFingerprintGenerator

def morgan_fp(sdf_file: str, out_npy: str, nbits:int=2048, radius:int=3):
    mfpgen = rdFingerprintGenerator.GetMorganGenerator(radius=radius,fpSize=nbits)
    suppl = Chem.SDMolSupplier(sdf_file)
    fps = list(mfpgen.GetFingerprint(m) for m in suppl if m is not None)
    npfps = np.array(fps).astype(np.uint8)
    np.save(out_npy, npfps)

if __name__ == '__main__':
    if len(sys.argv) != 5:
        print(f'Usage: {sys.argv[0]} <input sdf file> <output npy file> <Morgan FP radius> <Morgan FP bits>')
    sdf_path = sys.argv[1]
    npy = sys.argv[2]
    radius = int(sys.argv[3])
    nbits = int(sys.argv[4])
    morgan_fp(sdf_path, npy, nbits=nbits, radius=radius)

File: fingerprints.cpp
#include "cnpy.h"
#include <GraphMol/GraphMol.h>
#include <GraphMol/FileParsers/MolSupplier.h>
#include <GraphMol/Fingerprints/MorganFingerprints.h>
#include <cstdlib>


int main(int argc, char* argv[])
{
    // const unsigned int radius = 3;
    // const unsigned int nbits = 2048;

    if (argc != 5) {
        std::cerr << "Usage: " << argv[0] << " <input sdf file> <output npy file> <Morgan FP radius> <Morgan FP bits>\n";
        return 1;
    }

    std::string sdf_file = argv[1];
    std::string npy_file = argv[2];
    unsigned int radius = atoi(argv[3]);
    unsigned int nbits = atoi(argv[4]);
    std::unique_ptr<RDKit::ROMol> mol;
    std::unique_ptr<ExplicitBitVect> fp;
    bool takeOwnership = true;
    RDKit::SDMolSupplier mol_supplier(sdf_file, takeOwnership);
    std::vector<std::vector<uint8_t>> data;
    while (!mol_supplier.atEnd()) {
        mol.reset(mol_supplier.next());
        if (!mol) {continue;}
        fp.reset(RDKit::MorganFingerprints::getFingerprintAsBitVect(*mol, radius, nbits));
        std::vector<uint8_t> v;
        for (int i=0; i<nbits; ++i) {
            v.push_back(fp->getBit(i));
        }
        data.push_back(v);
    }
    cnpy::npy_save(npy_file, &data[0][0], {data.size(), nbits}, "w");
    return 0;
}
